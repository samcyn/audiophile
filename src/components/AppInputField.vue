<script lang="ts">
  // prevent inheritting attributes on root elements
  export default {
    inheritAttrs:
      false,
  };
</script>
<script setup lang="ts">
  import {
    defineProps,
    withDefaults,
    computed,
    InputHTMLAttributes,
    useAttrs,
  } from 'vue';

  // define component models
  interface AppInputFieldProps extends Omit<InputHTMLAttributes, 'class'> {
    variant?: 'contained' | 'text' | 'outlined';
    label?: string;
    name?: string;
    id?: string;
    value?: string;
    ariaLabel?: string;
    ariaLabelledby?: string;
    helperText?: string;
  }

  // set default props
  const props = withDefaults(defineProps<AppInputFieldProps>(), {
    variant: 'contained',
    label: undefined,
    name: '',
    id: '',
    value: '',
    ariaLabel: undefined,
    ariaLabelledby: undefined,
    helperText: undefined,
  });

  // pick attributes
  const attrs = useAttrs();

  const classComputed = computed<string>(() => {
    const variant = props.variant;
    let className = 'rounded-none transition ';
    if (variant === 'contained') {
      className += 'bg-orange-100 text-white hover:bg-orange-90';
    }
    if (variant === 'outlined') {
      className += 'bg-transparent hover:bg-black-100 border border-black-100 text-black-100 hover:text-white';
    }
    if (variant === 'text') {
      className += 'border-0 bg-none text-black-100 opacity-50 hover:text-orange-100 hover:opacity-100';
    }
    return `${className} ${attrs.class}`;
  });
</script>

<template>
  <div class="field relative">
    <label 
      :for="id" 
      class="label inline-flex items-center justify-between text-black-100 text-xs font-bold" 
      :class="{
        visuallyhidden: label === '',
      }"
    >
      <span class="text-inherit">
        {{ label }}
      </span>
      <span
        v-if="helperText" 
        class="text-inherit">
        {{ helperText }}
      </span>
    </label>
    <input 
      :aria-label="ariaLabel" 
      :aria-labelledby="ariaLabelledby" 
      type="text"
      class="
        input
        relative 
        block 
        w-full 
        border-0 
        rounded-lg
        bg-transparent
        ring-1 
        ring-inset 
        ring-grey-80 
        font-bold
        text-black-100 
        placeholder:text-black-100
        placeholder:opacity-40
        placeholder:font-bold
        focus:ring-1 
        focus:ring-inset 
        focus:ring-orange-100 
      "
      :name="name" 
      :id="id" 
      :value="value" 
      v-bind="$attrs" 
    />
  </div>
</template>

<style scoped>
  .label {
    font-style: normal;
    letter-spacing: -0.214286px;
    margin-bottom: 9px;
  }

  .input {
    font-style: normal;
    font-size: 14px;
    line-height: 19px;
    letter-spacing: -0.25px;
    mix-blend-mode: normal;
    padding: 17px 23px 18px;
    outline: 0;
  }

  .input::placeholder {
    font-style: normal;
    font-size: 14px;
    line-height: 19px;
    letter-spacing: -0.25px;
    mix-blend-mode: normal;
  }

  .input.error {
    border: 2px solid #cd2c2c;
  }

  .visuallyhidden {
    border: 0;
    clip: rect(0 0 0 0);
    height: 1px;
    margin: -1px;
    overflow: hidden;
    padding: 0;
    position: absolute;
    width: 1px;
  }
</style>
