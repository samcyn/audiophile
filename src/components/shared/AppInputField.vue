<script lang="ts">
// prevent inheritting attributes on root elements
export default {
  inheritAttrs: false,
};
</script>
<script setup lang="ts">
import { defineProps, withDefaults, InputHTMLAttributes, Events } from 'vue';

// define component models
interface AppInputFieldProps extends Omit<InputHTMLAttributes, 'class'> {
  label?: string;
  name?: string;
  id?: string;
  modelValue?: string;
  ariaLabel?: string;
  ariaLabelledby?: string;
  helperText?: string;
  hasError?: boolean;
  type?: string;
}

// set default props
withDefaults(defineProps<AppInputFieldProps>(), {
  label: undefined,
  name: '',
  id: '',
  ariaLabel: undefined,
  ariaLabelledby: undefined,
  helperText: undefined,
  modelValue: '',
  hasError: false,
  type: 'text',
});
const emit = defineEmits(['update:modelValue']);

const onInputChange = (event: Events['onInput']) => {
  const value = (event.target as HTMLInputElement).value;
  emit('update:modelValue', value);
};
</script>

<template>
  <div class="field relative">
    <label
      :for="id"
      class="label flex items-center justify-between w-full text-black-100 text-xs"
      :class="{
        visuallyhidden: label === '',
        'text-red': hasError,
      }"
    >
      <span class="text-inherit font-bold">
        {{ label }}
      </span>
      <span v-if="helperText" class="text-inherit font-medium">
        {{ helperText }}
      </span>
    </label>
    <input
      v-bind="$attrs"
      :aria-label="ariaLabel"
      :aria-labelledby="ariaLabelledby"
      :type="type"
      class="input relative block w-full border-0 rounded-lg bg-transparent ring-1 ring-inset ring-grey-80 font-bold text-black-100 placeholder:text-black-100 placeholder:opacity-40 placeholder:font-bold focus:ring-1 focus:ring-inset focus:ring-orange-100"
      :class="{
        'ring-red': hasError,
      }"
      :name="name"
      :id="id"
      :value="modelValue"
      @input="onInputChange"
    />
  </div>
</template>

<style scoped>
.label {
  font-style: normal;
  letter-spacing: -0.214286px;
  margin-bottom: 9px;
}

.input {
  font-style: normal;
  font-size: 14px;
  line-height: 19px;
  letter-spacing: -0.25px;
  mix-blend-mode: normal;
  padding: 18px 24px 19px;
  outline: 0;
}

.input::placeholder {
  font-style: normal;
  font-size: 14px;
  line-height: 19px;
  letter-spacing: -0.25px;
  mix-blend-mode: normal;
}

.input.error {
  border: 2px solid #cd2c2c;
}

.visuallyhidden {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}
</style>
