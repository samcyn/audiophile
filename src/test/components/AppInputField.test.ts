import { expect, test } from 'vitest';
import { mount } from '@vue/test-utils';
import AppInputField from '../../components/shared/AppInputField.vue';

test('mount AppInputField component', async () => {
  expect(AppInputField).toBeTruthy();

  const wrapper = mount(AppInputField, {
    props: {},
  });

  expect(wrapper.html()).toMatchSnapshot();
});
