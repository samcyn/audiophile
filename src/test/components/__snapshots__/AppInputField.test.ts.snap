// Vitest Snapshot v1

exports[`mount AppInputField component 1`] = `
"<div data-v-0eaff632=\\"\\" class=\\"field relative\\"><label data-v-0eaff632=\\"\\" for=\\"\\" class=\\"label flex items-center justify-between w-full text-black-100 text-xs\\"><span data-v-0eaff632=\\"\\" class=\\"text-inherit font-bold\\"></span>
    <!--v-if-->
  </label><input data-v-0eaff632=\\"\\" type=\\"text\\" class=\\"input relative block w-full border-0 rounded-lg bg-transparent ring-1 ring-inset ring-grey-80 font-bold text-black-100 placeholder:text-black-100 placeholder:opacity-40 placeholder:font-bold focus:ring-1 focus:ring-inset focus:ring-orange-100\\" name=\\"\\" id=\\"\\"></div>"
`;
